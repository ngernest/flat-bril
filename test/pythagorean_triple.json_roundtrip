{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "n",
            "n"
          ],
          "dest": "n_sq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "one"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "outer_loop"
        },
        {
          "args": [
            "one"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "inner_loop"
        },
        {
          "args": [
            "a",
            "a"
          ],
          "dest": "a_sq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "b",
            "b"
          ],
          "dest": "b_sq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a_sq",
            "b_sq"
          ],
          "dest": "sum",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "sum",
            "n_sq"
          ],
          "dest": "found",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "found"
          ],
          "funcs": [],
          "labels": [
            "found",
            "inner_continue"
          ],
          "op": "br"
        },
        {
          "label": "found"
        },
        {
          "args": [
            "b",
            "a"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "label": "inner_continue"
        },
        {
          "args": [
            "b",
            "one"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "b",
            "a"
          ],
          "dest": "done",
          "funcs": [],
          "labels": [],
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "done"
          ],
          "funcs": [],
          "labels": [
            "outer_continue",
            "inner_loop"
          ],
          "op": "br"
        },
        {
          "label": "outer_continue"
        },
        {
          "args": [
            "a",
            "one"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "a",
            "n"
          ],
          "dest": "done",
          "funcs": [],
          "labels": [],
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "done"
          ],
          "funcs": [],
          "labels": [
            "finish",
            "outer_loop"
          ],
          "op": "br"
        },
        {
          "label": "finish"
        }
      ],
      "name": "main"
    }
  ]
}
