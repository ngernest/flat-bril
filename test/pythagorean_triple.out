@main
	op: const		dest: one		type: int		value: 1	
	op: mul  		dest: n_sq		type: int		args: ["n", "n"]	
	op: id   		dest: a 		type: int		args: ["one"]	
.outer_loop
	op: id   		dest: b 		type: int		args: ["one"]	
.inner_loop
	op: mul  		dest: a_sq		type: int		args: ["a", "a"]	
	op: mul  		dest: b_sq		type: int		args: ["b", "b"]	
	op: add  		dest: sum		type: int		args: ["a_sq", "b_sq"]	
	op: eq   		dest: found		type: bool		args: ["sum", "n_sq"]	
	op: br   		args: ["found"]		labels: ["found", "inner_continue"]
.found
	op: print		args: ["b", "a"]	
.inner_continue
	op: add  		dest: b 		type: int		args: ["b", "one"]	
	op: ge   		dest: done		type: bool		args: ["b", "a"]	
	op: br   		args: ["done"]		labels: ["outer_continue", "inner_loop"]
.outer_continue
	op: add  		dest: a 		type: int		args: ["a", "one"]	
	op: ge   		dest: done		type: bool		args: ["a", "n"]	
	op: br   		args: ["done"]		labels: ["finish", "outer_loop"]
.finish
